File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章小朋友排队.cpp
//小朋友排队 lanqiaoOJ 222

#include <bits/stdc++.h>
using namespace std;
const int N = 1000010;
typedef long long LL;
int h[N], tree[N], k[N];        //h是身高，k是交换次数（逆序对数量）
int lowbit(int x) {  return x & -x; }
void update(int x, int d) {
    while(x <= N) {
        tree[x] += d;
        x += lowbit(x);
    }
}
int sum(int x) {
    int ans = 0;
    while(x > 0){
        ans += tree[x];
        x -= lowbit(x);
    }
    return ans;
}
int main() {
    int n; cin >> n;
    for (int i = 1; i <= n; i ++ ){
        cin >> h[i];
        h[i]++;          //h从1开始，不是从0开始
    }
    for (int i = 1; i <= n; i ++ ) {   //正序处理 逆序对，右边矮的
        k[i] = sum(N - 1) - sum(h[i]);
        update(h[i], 1);
    }
    memset(tree, 0, sizeof tree);
    for (int i = n; i; i-- ){         //倒序处理  逆序对，左边高的
        k[i] += sum(h[i] - 1);
        update(h[i], 1);
    }
    LL res = 0;
    for (int i = 1; i <= n; i ++ )    //最后把所有人的不愉快加起来
        res += (LL)k[i] * (k[i] + 1 ) / 2;
    cout << res;
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第6章小朋友排队.py
#小朋友排队  lanqiaoOJ 222

N = 1000010
def discretization(h):           #数组离散化
    temp=list(set(h))
    temp.sort()
    for i in range(len(h)):
        h[i]=temp.index(h[i])+1
def lowbit(x):
    return x&-x

def update(x,d):
    while x <= N:
        tree[x] += d
        x += lowbit(x)

def sum_(x):
    s = 0
    while(x>0):
        s+=tree[x]
        x-=lowbit(x)
    return s

n=int(input())
Hold=list(map(int,input().split()))
H=[0 for _ in range(N)]       #H是身高
for i in range(0,n):
    H[i+1] = Hold[i]+1
k=[0 for _ in range(N)]          #每个小朋友的最少交换次数. 也是逆序对数量
#discretization(H)               #不需要离散化
tree =[0 for _ in range(N)]
for i in range(1,n+1):           #正序处理 逆序对，右边矮的
    k[i] = sum_(N-1)- sum_(H[i])
    update(H[i],1)
tree =[0 for _ in range(N)]
for i in range(n,0,-1):          #倒序处理  逆序对，左边高的
    k[i]+=sum_(H[i]-1)
    update(H[i],1)
res=0
for i in range(1,n+1,1):
    res += int((1+k[i])*k[i]/2)
print(res)
